### PR TITLE
MRG+1: ENH: refactor beamformer channel picks with utils

### PR DESCRIPTION
**Pull request by britta-wstnr**
Moved some of the channel picking facility to ``utils``: 
``_check_info_inv`` (formerly ``_compute_beamformer._setup_picks``) picks channels based on the data / noise cov matrix and forward model, taking into account channels marked as bad in ``info`` and reference channels.
Next step for this PR is to write tests for this function.

Two additional functionalities have been added: check for compensation and dropping ref channels  during computation of spatial filter (will comment in code to mark those).

ping @agramfort

To Do:
- [x] add tests

### CODE REVIEW

File: mne/beamformer/_compute_beamformer.py [hunk 1]
```diff
@@ -166,6 +107,11 @@ def _prepare_beamformer_input(info, forward, label, picks, pick_ori,
         raise ValueError('Normal orientation can only be picked when a '
                          'forward operator with a surface-based source space '
                          'is used.')
+    # Check whether data and forward model have same compensation applied
+    if not get_current_comp(info) == get_current_comp(forward['info']):
+        raise ValueError('Data and forward model do not have same '
+                         'compensation applied.')
+
```
Comment (britta-wstnr) [inline mne/beamformer/_compute_beamformer.py]: @agramfort : this is new, checking whether compensation is identical between data and forward model

File: mne/beamformer/tests/test_lcmv.py [hunk 2]
```diff
@@ -648,5 +649,9 @@ def test_lcmv_ctf_comp():
     filters = make_lcmv(evoked.info, fwd, data_cov)
     assert 'weights' in filters
 
+    # test whether different compensations throw error
+    info_comp = evoked.info.copy()
+    set_current_comp(info_comp, 1)
+    pytest.raises(ValueError, make_lcmv, info_comp, fwd, data_cov)
```
Comment (britta-wstnr) [inline mne/beamformer/tests/test_lcmv.py]: @agramfort : this is the test for the compensation comparison

File: mne/utils/check.py [hunk 3]
```diff
@@ -330,3 +330,69 @@ def _check_if_nan(data, msg=" to be plotted"):
...
+    fwd_ch_names = forward['info']['ch_names']
+
+    # handle channels from forward model and info:
+    ch_names = _compare_ch_names(info['ch_names'], fwd_ch_names, info['bads'])
+
+    # make sure that no reference channels are left:
+    ref_chs = pick_types(info, meg=False, ref_meg=True)
+    ref_chs = [info['ch_names'][ch] for ch in ref_chs]
+    ch_names = [ch for ch in ch_names if ch not in ref_chs]
+
```
Comment (britta-wstnr) [inline mne/utils/check.py]: @agramfort : this code block is new as well, throwing out any reference channels that might still be in ``info`` for the beamformer. Now that we have this function in ``utils`` we probably need to reconsider whether this behavior is wanted by any inverse model?

Comment (larsoner) [reply mne/utils/check.py]: IIRC the forward does not contain the reference channels (only implicitly via compensation if it has been applied when computing the forward) so I don't think it should matter.

Comment (britta-wstnr) [reply mne/utils/check.py]: Yes, that is true, I acutally checked that. Talking to @agramfort, we thought about getting rid of them to ensure further computation doesn't go wrong in any case for the beamformer.

**Comment by agramfort** [comment]
+1 on this. For others, the idea is have common check functions that will be used across inverse methods that have data, a forward or filters or an inv and eventually some covs.

all inverse methods end up doing similar checks. Let's unify this.

sounds good?

**Comment by larsoner** [comment]
Yes! +1 for as much unification as possible across `minimum_norm`, `beamformer`, `inverse_sparse`, and `cov.py` for how this stuff is handled.

**Comment by britta-wstnr** [comment]
For test writing: ``test_check.py`` does not load any data (yet). ``_check_info_inv`` operates on ``info``, ``forward``, ``noise_cov``, and ``data_cov``. Is it worth loading real data for that, or should I rather build some fake dictionaries?

For the use of the new function anywhere but in the beamformer module I would be very glad about some help, as I don't know my way within those functions good enough yet to catch where it would be needed.

**Comment by agramfort** [comment]
no strong feeling about loading real data or not. We tend to use real data everywhere but maybe it would be interesting to consider having a way to mock more objects to speed up the tests by avoiding IO of complex fif files.

**Comment by britta-wstnr** [comment]
Based on ``test_lcmv.py``, the data covariance matrix would need to be estimated, but I guess that is no bottleneck either? If okay, I would add all this to ``test_check``

**Comment by britta-wstnr** [comment]
Thanks @larsoner misread the Travis output. Will fix and rebase!

**Comment by britta-wstnr** [comment]
@larsoner I think Travis timed out ...

**Comment by larsoner** [comment]
All better, ready for review/merge from your end?

File: mne/utils/tests/test_check.py [hunk 4]
```diff
@@ -36,3 +49,74 @@ def test_check_type_picks():
...
+    """Test checks for common channels across fwd model and cov matrices."""
+    epochs, data_cov, noise_cov, forward = _get_data()
+
+    # make sure same channel lists exist in data to make testing life easier
+    assert epochs.info['ch_names'] == data_cov.ch_names
+    assert epochs.info['ch_names'] == noise_cov.ch_names
+
+    # check whether bad channels get excluded from the channel selection
+    # info
+    info_bads = deepcopy(epochs.info)
```
Comment (agramfort) [inline mne/utils/tests/test_check.py]: use epochs.info.copy()

File: mne/utils/tests/test_check.py [hunk 5]
```diff
@@ -36,3 +49,74 @@ def test_check_type_picks():
...
+    info_bads['bads'] = info_bads['ch_names'][1:3]  # include two bad channels
+    picks = _check_info_inv(info_bads, forward, noise_cov=noise_cov)
+    assert [1, 2] not in picks
+    # covariance matrix
+    data_cov_bads = deepcopy(data_cov)
```
Comment (agramfort) [inline mne/utils/tests/test_check.py]: data_cov.copy()

File: mne/utils/tests/test_check.py [hunk 6]
```diff
@@ -36,3 +49,74 @@ def test_check_type_picks():
...
+    data_cov_bads['bads'] = data_cov_bads.ch_names[0]
+    picks = _check_info_inv(epochs.info, forward, data_cov=data_cov_bads)
+    assert 0 not in picks
+    # noise covariance matrix
+    noise_cov_bads = deepcopy(noise_cov)
```
Comment (agramfort) [inline mne/utils/tests/test_check.py]: idem

**Comment by britta-wstnr** [comment]
Addressed @agramfort 's comments, will get to @larsoner 's a bit later. Thanks guys!

**Comment by britta-wstnr** [comment]
ping @larsoner : I need help with CIs again ... I don't think it is related to what I did? Should I rebase?

**Comment by larsoner** [comment]
Ignore errors related to physionet download. This will require checking Travis logs. Or wait an hour until #5932 is in, and I can restart Travis (which builds a merged version of the PR, so will get the fixes automatically on build restart).

**Comment by massich** [comment]
@larsoner 	the failing download is actually this one https://travis-ci.org/mne-tools/mne-python/jobs/490227429#L2490 it does it sometimes. 

but yes, right now the CIs are broken. My fault.

**Comment by britta-wstnr** [comment]
Okay, thanks -- then let's see what Travis says later!

**Comment by larsoner** [comment]
Travis isn't building a merged version (I forgot it still uses the old `.travis.yml`, and merges the rest of the code), you can rebase, or just ignore the 3.7 build for now.

File: mne/utils/__init__.py [hunk 7]
```diff
@@ -1,6 +1,6 @@
 # # # WARNING # # #
-# This list must also be updated in doc/_templates/class.rst if it is
-# changed here!
+# This list must also be updated in doc/_templates/autosummary/class.rst if it
+# is changed here!
```
Comment (britta-wstnr) [inline mne/utils/__init__.py]: @larsoner I think the location of ```class.rst``` has changed?

**Comment by agramfort** [comment]
we could merge this but I would suggest to see if the new _check_info_inv can be shared among all inverse solvers. @larsoner can you have a look?

File: mne/utils/__init__.py (hunk 7 above)
Comment (larsoner) [reply mne/utils/__init__.py]: Indeed it has, thanks for the fix

**Comment by larsoner** [comment]
@agramfort @britta-wstnr I started looking into it. There is a lot of overlap between:

1. `mne.minimum_norm.inverse._prepare_forward` plus the depth-prior-computing parts of `make_inverse_operator`, and
2. `mne.beamformer._compute_beamformer._prepare_beamformer_input` plus `_check_info_inv` (which could actually be refactored to live in `_prepare_beamformer_input`).

These two code paths seem to do in principle the same things:

1. check for channel naming consistency
2. do channel picking of the gain matrix and cov matrix
3. check comps
4. compute a depth prior

The best thing to do is probably to merge these things, probably by modifying `_prepare_forward` as necessary. Let me know if you think it makes sense to do here, which could delay merge for some days or weeks, or if you'd rather see a separate PR.

**Comment by agramfort** [comment]
that's what I feared... no strong feeling. Do you think it would take time
to fix?

>

**Comment by larsoner** [comment]
I didn't think `rank` would take very long but it did. Hopefully only an hour or so, but who knows. Regardless, we should do it.

I'll try to add it here (won't push until it's close) and if it takes too long we can just merge.

**Comment by larsoner** [comment]
I think we should merge this as-is, and I'll tackle unification separately. It's going to need to be a multi-step process, and there's no need to hold this up in the meantime.

**Comment by massich** [comment]
ok to merge as is, and refactor in #5947

**Pull request merged**